### ISSUE TITLE
get.trait.data.pft erroneously skipping updating trait records

### ISSUE THREAD

**Issue by bcow** [PecanProject/pecan#2422]
This happens twice in the code: 
I'll come back and link these to code lines once the PR where I've moved around functions goes through.

# First:

```{r}
  # Set forceupdate FALSE if it's a string (backwards compatible with 'AUTO' flag used in the past)
  if (!is.logical(forceupdate)) {
    forceupdate <- FALSE
  }
```

This is a problem because if you use the `<update>` flag in the `pecan.xml` as I do sometimes to force update the meta analysis, then `forceupdate = "TRUE"` will still come though as a string. Just double checking that I'm not missing something here, and this is just a simple fix and I can just add in 

```
 forceupdate <- try(as.logical(forceupdate))
 if (!is.logical(forceupdate)) {
    forceupdate <- FALSE
  }
```

# Second

Further down in the same function, if force update is not TRUE, then an automated check goes through to see if an update is needed. 

```
current_traits <- dplyr::bind_rows(trait.data.check, .id = "trait") %>%
              dplyr::select(-mean, -stat) 
existing_traits <- dplyr::bind_rows(existing_trait_data, .id = "trait") %>%
              dplyr::select(-mean, -stat) # %>% mutate(statname = as.character(statname))
diff_traits <- symmetric_setdiff(current_traits, existing_traits)
if (nrow(diff_traits) > 0) {
             # need to update
}
```

Even when the two data frames ("current" and "existing") are different, `symmetric_setdiff` returns an empty data frame. I'm currently testing with two data frames that not only have different entries but they also have a different number of rows and I still get an empty result.
On the other hand, regular `setdiff` returns almost all the entries. The data frames I am testing with should only differ by about nine rows. So neither of the two functions seem to be getting me what I need. 

I arrived at this solution:

```            
test <- dplyr::all_equal(existing_traits, current_traits, ignore_col_order = TRUE,
             ignore_row_order = TRUE, convert = TRUE)

if (isTRUE(test)) {
             # need to update
}
``` 

Does `all_equal` seem like a sufficient solution?

**Comment by mdietze** [comment]
I don't know if this was dropped at some point, but there was a time where the trait MA update code could be set to "AUTO", not just T/F, so that the use could ask PEcAn to check whether the trait data needed to be updated, rather than forcing it to update or forcing it to not update. If that's still supported I worry that `forceupdate <- try(as.logical(forceupdate))` will change "AUTO" into a `try` error message.

**Comment by bcow** [comment]
This is the output I get when running the code setting `forceupdate = "AUTO"`
It seems to work. That being said, I agree, I thought that the output would also be a `try` error message. Perhaps this is not the right kind of solution. 

```
> forceupdate <- "AUTO"
> try(as.logical(forceupdate))
[1] NA
> if (!is.logical(forceupdate)) {
+   forceupdate <- FALSE
+ }
> forceupdate
[1] FALSE
```

**Comment by bcow** [comment]
Yes I thought that the `try` output was working the way I wanted but checking it again - I don't think it's what I want to use.

**Comment by bcow** [comment]
Perhaps 

```
forceupdate <- isTRUE(as.logical(forceupdate))
```

Then 

```
> isTRUE(as.logical("TRUE"))
[1] TRUE
> isTRUE(as.logical("FALSE"))
[1] FALSE
```

And 
```
> isTRUE(as.logical("AUTO"))
[1] FALSE
```

**Comment by para2x** [comment]
@bcow I told you this in a pm but for later reference for other people I think this could be a solution for your second issue : 
```
match <- existing_traits %>%
  purrr::map2(current_traits, all.equal.list)
````

**Comment by bcow** [comment]
I'm not closing this just yet until I've pushed a PR with fixes to the first problem and I've tested `symmetric_setdiff` more.

**Comment by KristinaRiemer** [comment]
@bcow did you ever submit a PR for this? I'm dealing with this same problem now, of not being able to set update to true.

**Comment by ashiklom** [comment]
@KristinaRiemer Did #2755 resolve this?

**Comment by Aariq** [comment]
Looks like this was fixed in #2755
